`spack.config.override()` prepends to lists rather than replacing

Given the case where a list configuration item is specified in YAML (`config.yaml`, say):
```YAML
...
extensions:
  - /my/path/to/extension
```
and a test overrides that item, _viz_:
```Python
    with spack.config.override('config:extensions', ['/my/path/to/extension-1', '/my/path/to/extension-2']):
        ...
```
Then the actual value of `config:extensions` inside the overridden context as returned by `spack.config.get()` is:
```Python
['/my/path/to/extension-1', '/my/path/to/extension-2', '/my/path/to/extension']
```
rather than the expected:
```Python
['/my/path/to/extension-1', '/my/path/to/extension-2']
```

I would expect higher priority scopes to replace lists rather than prepend to them, at least as a default behavior. Is this intended, and what is the use case?

@tldahlgren 's suggestion results in the same `KeyError`. Adding a `:` or `::` to a `spack.config.get()` call results in the error:
```console
E           AttributeError: 'CommentedSeq' object has no attribute 'get'
```
in both cases.

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -280,6 +280,7 @@ class InternalConfigScope(ConfigScope):
         self.sections = syaml.syaml_dict()
 
         if data:
+            data = _process_dict(data)
             for section in data:
                 dsec = data[section]
                 validate({section: dsec}, section_schemas[section])
@@ -505,14 +506,14 @@ class Configuration(object):
 
         Accepts the path syntax described in ``get()``.
         """
-        section, _, rest = path.partition(':')
+        parts = _process_path(path)
+        section = parts.pop(0)
 
-        if not rest:
+        if not parts:
             self.update_config(section, value, scope=scope)
         else:
             section_data = self.get_config(section, scope=scope)
 
-            parts = rest.split(':')
             data = section_data
             while len(parts) > 1:
                 key = parts.pop(0)
@@ -612,7 +613,7 @@ def _config():
     """Singleton Configuration instance.
 
     This constructs one instance associated with this module and returns
-    it. It is bundled inside a function so that configuratoin can be
+    it. It is bundled inside a function so that configuration can be
     initialized lazily.
 
     Return:
@@ -752,7 +753,7 @@ def _mark_internal(data, name):
     return d
 
 
-def _merge_yaml(dest, source):
+def _merge_yaml(dest, source, override=False):
     """Merges source into dest; entries in source take precedence over dest.
 
     This routine may modify dest and should be assigned to dest, in
@@ -764,57 +765,93 @@ def _merge_yaml(dest, source):
     with `::` instead of `:`, and the key will override that of the
     parent instead of merging.
 
+    If override is True, then source will be copied to dest unconditionally.
     """
     def they_are(t):
         return isinstance(dest, t) and isinstance(source, t)
 
-    # If both are None, handle specially and return None.
-    if source is None and dest is None:
+    if source is None:
         return None
 
-    # If source is None, overwrite with source.
-    elif source is None:
-        return None
+    if not override:
+        # Source list is prepended (for precedence)
+        if they_are(list):
+            dest[:] = source + [x for x in dest if x not in source]
+            return dest
 
-    # Source list is prepended (for precedence)
-    if they_are(list):
-        dest[:] = source + [x for x in dest if x not in source]
-        return dest
+        # Source dict is merged into dest.
+        elif they_are(dict):
+            # track keys for marking
+            key_marks = {}
 
-    # Source dict is merged into dest.
-    elif they_are(dict):
-        # track keys for marking
-        key_marks = {}
+            for sk, sv in iteritems(source):
+                if _override(sk) or sk not in dest:
+                    # if sk ended with ::, or if it's new, completely override
+                    dest[sk] = copy.copy(sv)
+                else:
+                    # otherwise, merge the YAML
+                    dest[sk] = _merge_yaml(dest[sk], source[sk], _override(sk))
 
-        for sk, sv in iteritems(source):
-            if _override(sk) or sk not in dest:
-                # if sk ended with ::, or if it's new, completely override
-                dest[sk] = copy.copy(sv)
-            else:
-                # otherwise, merge the YAML
-                dest[sk] = _merge_yaml(dest[sk], source[sk])
+                # this seems unintuitive, but see below. We need this because
+                # Python dicts do not overwrite keys on insert, and we want
+                # to copy mark information on source keys to dest.
+                key_marks[sk] = sk
 
-            # this seems unintuitive, but see below. We need this because
-            # Python dicts do not overwrite keys on insert, and we want
-            # to copy mark information on source keys to dest.
-            key_marks[sk] = sk
+            # ensure that keys are marked in the destination. The
+            # key_marks dict ensures we can get the actual source key
+            # objects from dest keys
+            for dk in list(dest.keys()):
+                if dk in key_marks and syaml.markable(dk):
+                    syaml.mark(dk, key_marks[dk])
+                elif dk in key_marks:
+                    # The destination key may not be markable if it was derived
+                    # from a schema default. In this case replace the key.
+                    val = dest.pop(dk)
+                    dest[key_marks[dk]] = val
 
-        # ensure that keys are marked in the destination.  the key_marks dict
-        # ensures we can get the actual source key objects from dest keys
-        for dk in list(dest.keys()):
-            if dk in key_marks and syaml.markable(dk):
-                syaml.mark(dk, key_marks[dk])
-            elif dk in key_marks:
-                # The destination key may not be markable if it was derived
-                # from a schema default. In this case replace the key.
-                val = dest.pop(dk)
-                dest[key_marks[dk]] = val
+            return dest
 
-        return dest
+    # If we reach here, overwrite with a copy of the source value.
+    return copy.copy(source)
 
-    # In any other case, overwrite with a copy of the source value.
-    else:
-        return copy.copy(source)
+
+#
+# Process a path argument to config.set() that may contain overrides ('::' or
+# trailing ':')
+#
+def _process_path(path):
+    result = []
+    if path.startswith(':'):
+        raise syaml.SpackYAMLError("Illegal leading `:' in path `{0}'".
+                                   format(path), '')
+    while path:
+        front, _, path = path.partition(':')
+        if (_ and not path) or path.startswith(':'):
+            path = path.lstrip(':')
+            front = syaml.syaml_str(front)
+            front.override = True
+        result.append(front)
+    return result
+
+
+#
+# Process a dict whose keys may describe overrides (trailing ':')
+#
+def _process_dict(data):
+    result = {}
+    for sk, sv in iteritems(data):
+        if sk.endswith(':'):
+            key = syaml.syaml_str(sk[:-1])
+            key.override = True
+        else:
+            key = sk
+
+        if isinstance(sv, dict):
+            result[key] = _process_dict(sv)
+        else:
+            result[key] = copy.copy(sv)
+
+    return result
 
 
 #

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -825,8 +825,8 @@ def _process_config_path(path):
         raise syaml.SpackYAMLError("Illegal leading `:' in path `{0}'".
                                    format(path), '')
     while path:
-        front, _, path = path.partition(':')
-        if (_ and not path) or path.startswith(':'):
+        front, sep, path = path.partition(':')
+        if (sep and not path) or path.startswith(':'):
             path = path.lstrip(':')
             front = syaml.syaml_str(front)
             front.override = True

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -783,8 +783,6 @@ def _merge_yaml(dest, source):
     Config file authors can optionally end any attribute in a dict
     with `::` instead of `:`, and the key will override that of the
     parent instead of merging.
-
-    If override is True, then source will be copied to dest unconditionally.
     """
     def they_are(t):
         return isinstance(dest, t) and isinstance(source, t)

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -772,7 +772,7 @@ def _mark_internal(data, name):
     return d
 
 
-def _merge_yaml(dest, source, override=False):
+def _merge_yaml(dest, source):
     """Merges source into dest; entries in source take precedence over dest.
 
     This routine may modify dest and should be assigned to dest, in
@@ -792,45 +792,45 @@ def _merge_yaml(dest, source, override=False):
     if source is None:
         return None
 
-    if not override:
-        # Source list is prepended (for precedence)
-        if they_are(list):
-            dest[:] = source + [x for x in dest if x not in source]
-            return dest
+    # Source list is prepended (for precedence)
+    if they_are(list):
+        dest[:] = source + [x for x in dest if x not in source]
+        return dest
 
-        # Source dict is merged into dest.
-        elif they_are(dict):
-            # track keys for marking
-            key_marks = {}
+    # Source dict is merged into dest.
+    elif they_are(dict):
+        # track keys for marking
+        key_marks = {}
 
-            for sk, sv in iteritems(source):
-                if _override(sk) or sk not in dest:
-                    # if sk ended with ::, or if it's new, completely override
-                    dest[sk] = copy.copy(sv)
-                else:
-                    # otherwise, merge the YAML
-                    dest[sk] = _merge_yaml(dest[sk], source[sk], _override(sk))
+        for sk, sv in iteritems(source):
+            if _override(sk) or sk not in dest:
+                # if sk ended with ::, or if it's new, completely override
+                dest[sk] = copy.copy(sv)
+            else:
+                # otherwise, merge the YAML
+                dest[sk] = _merge_yaml(dest[sk], source[sk])
 
-                # this seems unintuitive, but see below. We need this because
-                # Python dicts do not overwrite keys on insert, and we want
-                # to copy mark information on source keys to dest.
-                key_marks[sk] = sk
+            # this seems unintuitive, but see below. We need this because
+            # Python dicts do not overwrite keys on insert, and we want
+            # to copy mark information on source keys to dest.
+            key_marks[sk] = sk
 
-            # ensure that keys are marked in the destination. The
-            # key_marks dict ensures we can get the actual source key
-            # objects from dest keys
-            for dk in list(dest.keys()):
-                if dk in key_marks and syaml.markable(dk):
-                    syaml.mark(dk, key_marks[dk])
-                elif dk in key_marks:
-                    # The destination key may not be markable if it was derived
-                    # from a schema default. In this case replace the key.
-                    val = dest.pop(dk)
-                    dest[key_marks[dk]] = val
+        # ensure that keys are marked in the destination. The
+        # key_marks dict ensures we can get the actual source key
+        # objects from dest keys
+        for dk in list(dest.keys()):
+            if dk in key_marks and syaml.markable(dk):
+                syaml.mark(dk, key_marks[dk])
+            elif dk in key_marks:
+                # The destination key may not be markable if it was derived
+                # from a schema default. In this case replace the key.
+                val = dest.pop(dk)
+                dest[key_marks[dk]] = val
 
-            return dest
+        return dest
 
-    # If we reach here, overwrite with a copy of the source value.
+    # If we reach here source and dest are either different types or are
+    # not both lists or dicts: replace with source.
     return copy.copy(source)
 
 

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -280,7 +280,7 @@ class InternalConfigScope(ConfigScope):
         self.sections = syaml.syaml_dict()
 
         if data:
-            data = _process_dict(data)
+            data = InternalConfigScope._process_dict_keyname_overrides(data)
             for section in data:
                 dsec = data[section]
                 validate({section: dsec}, section_schemas[section])
@@ -306,6 +306,25 @@ class InternalConfigScope(ConfigScope):
 
     def __repr__(self):
         return '<InternalConfigScope: %s>' % self.name
+
+    @staticmethod
+    def _process_dict_keyname_overrides(data):
+        """Turn a trailing `:' in a key name into an override attribute."""
+        result = {}
+        for sk, sv in iteritems(data):
+            if sk.endswith(':'):
+                key = syaml.syaml_str(sk[:-1])
+                key.override = True
+            else:
+                key = sk
+
+            if isinstance(sv, dict):
+                result[key]\
+                    = InternalConfigScope._process_dict_keyname_overrides(sv)
+            else:
+                result[key] = copy.copy(sv)
+
+        return result
 
 
 class Configuration(object):
@@ -831,26 +850,6 @@ def _process_config_path(path):
             front = syaml.syaml_str(front)
             front.override = True
         result.append(front)
-    return result
-
-
-#
-# Process a dict whose keys may describe overrides (trailing ':')
-#
-def _process_dict(data):
-    result = {}
-    for sk, sv in iteritems(data):
-        if sk.endswith(':'):
-            key = syaml.syaml_str(sk[:-1])
-            key.override = True
-        else:
-            key = sk
-
-        if isinstance(sv, dict):
-            result[key] = _process_dict(sv)
-        else:
-            result[key] = copy.copy(sv)
-
     return result
 
 

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -787,6 +787,7 @@ def _merge_yaml(dest, source):
     def they_are(t):
         return isinstance(dest, t) and isinstance(source, t)
 
+    # If source is None, overwrite with source.
     if source is None:
         return None
 

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -506,7 +506,7 @@ class Configuration(object):
 
         Accepts the path syntax described in ``get()``.
         """
-        parts = _process_path(path)
+        parts = _process_config_path(path)
         section = parts.pop(0)
 
         if not parts:
@@ -819,7 +819,7 @@ def _merge_yaml(dest, source, override=False):
 # Process a path argument to config.set() that may contain overrides ('::' or
 # trailing ':')
 #
-def _process_path(path):
+def _process_config_path(path):
     result = []
     if path.startswith(':'):
         raise syaml.SpackYAMLError("Illegal leading `:' in path `{0}'".

--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -843,12 +843,18 @@ def _process_config_path(path):
     if path.startswith(':'):
         raise syaml.SpackYAMLError("Illegal leading `:' in path `{0}'".
                                    format(path), '')
+    seen_override_in_path = False
     while path:
         front, sep, path = path.partition(':')
         if (sep and not path) or path.startswith(':'):
+            if seen_override_in_path:
+                raise syaml.SpackYAMLError("Meaningless second override"
+                                           " indicator `::' in path `{0}'".
+                                           format(path), '')
             path = path.lstrip(':')
             front = syaml.syaml_str(front)
             front.override = True
+            seen_override_in_path = True
         result.append(front)
     return result
 

--- a/lib/spack/spack/test/config.py
+++ b/lib/spack/spack/test/config.py
@@ -881,7 +881,7 @@ def test_internal_config_section_override(mock_low_high_config,
                                             'build_stage': wanted_list
                                         }
                                     }))
-    assert wanted_list == mock_low_high_config.get('config:build_stage')
+    assert mock_low_high_config.get('config:build_stage') == wanted_list
 
 
 def test_internal_config_dict_override(mock_low_high_config,
@@ -890,7 +890,7 @@ def test_internal_config_dict_override(mock_low_high_config,
     wanted_dict = config_override_dict['config']['info:']
     mock_low_high_config.push_scope(spack.config.InternalConfigScope
                                     ('high', config_override_dict))
-    assert wanted_dict == mock_low_high_config.get('config:info')
+    assert mock_low_high_config.get('config:info') == wanted_dict
 
 
 def test_internal_config_list_override(mock_low_high_config,
@@ -899,14 +899,14 @@ def test_internal_config_list_override(mock_low_high_config,
     wanted_list = config_override_list['config']['build_stage:']
     mock_low_high_config.push_scope(spack.config.InternalConfigScope
                                     ('high', config_override_list))
-    assert wanted_list == mock_low_high_config.get('config:build_stage')
+    assert mock_low_high_config.get('config:build_stage') == wanted_list
 
 
 def test_set_section_override(mock_low_high_config, write_config_file):
     write_config_file('config', config_merge_list, 'low')
     wanted_list = config_override_list['config']['build_stage:']
     with spack.config.override('config::build_stage', wanted_list):
-        assert wanted_list == mock_low_high_config.get('config:build_stage')
+        assert mock_low_high_config.get('config:build_stage') == wanted_list
     assert config_merge_list['config']['build_stage'] == \
         mock_low_high_config.get('config:build_stage')
 

--- a/lib/spack/spack/test/config.py
+++ b/lib/spack/spack/test/config.py
@@ -46,7 +46,19 @@ config_merge_list = {
 
 config_override_list = {
     'config': {
-        'build_stage:': ['patha', 'pathb']}}
+        'build_stage:': ['pathd', 'pathe']}}
+
+config_merge_dict = {
+    'config': {
+        'info': {
+            'a': 3,
+            'b': 4}}}
+
+config_override_dict = {
+    'config': {
+        'info:': {
+            'a': 7,
+            'c': 9}}}
 
 
 @pytest.fixture()
@@ -382,7 +394,7 @@ def test_read_config_override_list(mock_low_high_config, write_config_file):
     write_config_file('config', config_override_list, 'high')
     assert spack.config.get('config') == {
         'install_tree': 'install_tree_path',
-        'build_stage': ['patha', 'pathb']
+        'build_stage': config_override_list['config']['build_stage:']
     }
 
 
@@ -857,3 +869,67 @@ def test_dotkit_in_config_does_not_raise(
     # we throw a a deprecation warning without raising
     assert '_sp_sys_type' in captured[0]  # stdout
     assert 'Warning' in captured[1]  # stderr
+
+
+def test_internal_config_section_override(mock_low_high_config,
+                                          write_config_file):
+    write_config_file('config', config_merge_list, 'low')
+    wanted_list = config_override_list['config']['build_stage:']
+    mock_low_high_config.push_scope(spack.config.InternalConfigScope
+                                    ('high', {
+                                        'config:': {
+                                            'build_stage': wanted_list
+                                        }
+                                    }))
+    assert wanted_list == mock_low_high_config.get('config:build_stage')
+
+
+def test_internal_config_dict_override(mock_low_high_config,
+                                       write_config_file):
+    write_config_file('config', config_merge_dict, 'low')
+    wanted_dict = config_override_dict['config']['info:']
+    mock_low_high_config.push_scope(spack.config.InternalConfigScope
+                                    ('high', config_override_dict))
+    assert wanted_dict == mock_low_high_config.get('config:info')
+
+
+def test_internal_config_list_override(mock_low_high_config,
+                                       write_config_file):
+    write_config_file('config', config_merge_list, 'low')
+    wanted_list = config_override_list['config']['build_stage:']
+    mock_low_high_config.push_scope(spack.config.InternalConfigScope
+                                    ('high', config_override_list))
+    assert wanted_list == mock_low_high_config.get('config:build_stage')
+
+
+def test_set_section_override(mock_low_high_config, write_config_file):
+    write_config_file('config', config_merge_list, 'low')
+    wanted_list = config_override_list['config']['build_stage:']
+    with spack.config.override('config::build_stage', wanted_list):
+        assert wanted_list == mock_low_high_config.get('config:build_stage')
+    assert config_merge_list['config']['build_stage'] == \
+        mock_low_high_config.get('config:build_stage')
+
+
+def test_set_list_override(mock_low_high_config, write_config_file):
+    write_config_file('config', config_merge_list, 'low')
+    wanted_list = config_override_list['config']['build_stage:']
+    with spack.config.override('config:build_stage:', wanted_list):
+        assert wanted_list == mock_low_high_config.get('config:build_stage')
+    assert config_merge_list['config']['build_stage'] == \
+        mock_low_high_config.get('config:build_stage')
+
+
+def test_set_dict_override(mock_low_high_config, write_config_file):
+    write_config_file('config', config_merge_dict, 'low')
+    wanted_dict = config_override_dict['config']['info:']
+    with spack.config.override('config:info:', wanted_dict):
+        assert wanted_dict == mock_low_high_config.get('config:info')
+    assert config_merge_dict['config']['info'] == \
+        mock_low_high_config.get('config:info')
+
+
+def test_set_bad_path(config):
+    with pytest.raises(syaml.SpackYAMLError):
+        with spack.config.override(':bad:path', ''):
+            pass

--- a/lib/spack/spack/test/config.py
+++ b/lib/spack/spack/test/config.py
@@ -930,6 +930,13 @@ def test_set_dict_override(mock_low_high_config, write_config_file):
 
 
 def test_set_bad_path(config):
-    with pytest.raises(syaml.SpackYAMLError):
+    with pytest.raises(syaml.SpackYAMLError, match='Illegal leading'):
         with spack.config.override(':bad:path', ''):
+            pass
+
+
+def test_bad_path_double_override(config):
+    with pytest.raises(syaml.SpackYAMLError,
+                       match='Meaningless second override'):
+        with spack.config.override('bad::double:override::directive', ''):
             pass

